prune/prune.go

```diff
@@ -16,184 +16,209 @@ package prune
-			if err != nil {
-				return err
-			}
-			log.V(1).Info("pods ", "count", len(resourceList))
-		} else if config.Resources[i].Kind == JobKind {
-			resourceList, err = config.getCompletedJobs(ctx)
-			if err != nil {
-				return err
-			}
-			log.V(1).Info("jobs ", "count", len(resourceList))
-		}
+	// gvk is the type of objects to prune.
+	// It defaults to Pod
+	gvk schema.GroupVersionKind
 
-		var resourcesToRemove []ResourceInfo
-
-		switch config.Strategy.Mode {
-		case MaxAgeStrategy:
-			resourcesToRemove, err = pruneByMaxAge(ctx, config, resourceList)
-		case MaxCountStrategy:
-			resourcesToRemove, err = pruneByMaxCount(ctx, config, resourceList)
-		case CustomStrategy:
-			resourcesToRemove, err = config.CustomStrategy(ctx, config, resourceList)
-		default:
-			return fmt.Errorf("unknown strategy")
-		}
-		if err != nil {
-			return err
-		}
+	// strategy is the function used to determine a list of resources that are pruneable
+	strategy StrategyFunc
 
-		err = config.removeResources(ctx, resourcesToRemove)
-		if err != nil {
-			return err
-		}
-	}
+	// labels is a map of the labels to use for label matching when looking for resources
+	labels map[string]string
 
-	log.V(1).Info("Prune completed")
+	// namespace is the namespace to use when looking for resources
+	namespace string
 
-	return nil
+	// logger is the logger to use when running pruning functionality
+	logger logr.Logger
 }
 
-// containsString checks if a string is present in a slice
-func containsString(s []string, str string) bool {
-	for _, v := range s {
-		if v == str {
-			return true
-		}
+// Unprunable indicates that it is not allowed to prune a specific object.
+type Unprunable struct {
+	Obj    *client.Object
+	Reason string
+}
+
+// Error returns a string representation of an `Unprunable` error.
+func (e *Unprunable) Error() string {
+	return fmt.Sprintf("unable to prune %s: %s", client.ObjectKeyFromObject(*e.Obj), e.Reason)
+}
+
+// StrategyFunc takes a list of resources and returns the subset to prune.
+type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)
+
+// IsPrunableFunc is a function that checks the data of an object to see whether or not it is safe to prune it.
+// It should return `nil` if it is safe to prune, `Unprunable` if it is unsafe, or another error.
+// It should safely assert the object is the expected type, otherwise it might panic.
+type IsPrunableFunc func(obj client.Object, logger logr.Logger) error
+
+// PrunerOption configures the pruner.
+type PrunerOption func(p *Pruner)
+
+// WithNamespace can be used to set the Namespace field when configuring a Pruner
+func WithNamespace(namespace string) PrunerOption {
+	return func(p *Pruner) {
+		p.namespace = namespace
 	}
+}
 
-	return false
+// WithLogger can be used to set the Logger field when configuring a Pruner
+func WithLogger(logger logr.Logger) PrunerOption {
+	return func(p *Pruner) {
+		p.logger = logger
+	}
```

Comment:
After talking with @everettraven we came to the conclusion that we can and should remove the logger from the pruner configuration. None of the internal pruning logic actually uses the logger so we can avoid coupling the library to any logger implementation. Users who want logging can implement their own is-pruneable and strategy functions that do whatever logging they want. They can even wrap the existing is-pruneable functions to add logging:

``` go
func myPodIsPruneable(obj client.Client) error {
    if err := prune.DefaultPodIsPruneable(obj); if prune.IsUnprunable(err) {
        // log some fun stuff
    } else if err != nil {
        // log some less-fun stuff
        return err
    }
    
    return nil
}
```

Reply:
All logging implementation is removed with commit https://github.com/operator-framework/operator-lib/pull/105/commits/be7adfb415a270d4b5483d58e3d3a0abfc157934